### PR TITLE
MediatR Guide + sample

### PR DESCRIPTION
**Pull request by dukesteen**
This PR introduces a guide to adding MediatR to your Discord.Net bot. it goes over adding MediatR to your DI container, creating an event listener, registering said event listener inside your DI container, notifications, and notification handlers.

### CODE REVIEW

File: docs/guides/other_libs/samples/MediatrCreatingMessageNotification.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// MessageReceivedNotification.cs
+
+using Discord.WebSocket;
+using MediatR;
+
+namespace Viscoin.Bot.Infrastructure.Messages;
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrCreatingMessageNotification.cs]: `namespace MediatRSample.Notifications;`

File: docs/guides/other_libs/mediatr.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+## Testing
+
+To test if we have successfully implemented MediatR, we can start up the bot and send a message to a server the bot is in. It should print out the message we defined earlier in our `MessageReceivedHandler`.
+
+![MediatR output](images/mediatr_output.png)
+
+## Adding more event types
+
+To add more event types you can follow these steps
```
Comment (csmir) [inline docs/guides/other_libs/mediatr.md]: 'steps:'

File: docs/guides/other_libs/mediatr.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+1. Create a new notification class for the event. it should contain all of the parameters that the event would send. (Ex: the `MessageReceived` event takes one `SocketMessage` as an argument. The notification class should also map this argument)
+2. Register the event in your `DiscordEventListener` class
```
Comment (csmir) [inline docs/guides/other_libs/mediatr.md]: 'class.'

File: docs/guides/other_libs/samples/MediatrDiscordEventListener.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    private IMediator Mediator
+    {
+        get
+        {
+            var scope = _serviceScope.CreateScope();
+            return scope.ServiceProvider.GetRequiredService<IMediator>();
+        }
+    }
+
+    public Task StartAsync()
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrDiscordEventListener.cs]: `public async Task StartAsync()` considering this call is awaited in `Program.cs`

File: docs/guides/other_libs/samples/MediatrDiscordEventListener.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        _client.MessageReceived += OnMessageReceivedAsync;
+
+        return Task.CompletedTask;
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrDiscordEventListener.cs]: `await Task.CompletedTask;`

File: docs/guides/other_libs/samples/MediatrMessageReceivedHandler.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// MessageReceivedHandler.cs
+
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrMessageReceivedHandler.cs:2]: `using System;`

File: docs/guides/other_libs/samples/MediatrDiscordEventListener.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private Task OnMessageReceivedAsync(SocketMessage arg)
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrDiscordEventListener.cs:42]: if `Mediator.Publish` is a `Task`, we should not block the gateway task here, I do not know if disposal is already done within the publishment?

File: docs/guides/other_libs/samples/MediatrDiscordEventListener.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// DiscordEventListener.cs
+
```
Comment (csmir) [inline docs/guides/other_libs/samples/MediatrDiscordEventListener.cs:2]: `using System.Threading;`
`using System.Threading.Tasks;`

File: docs/guides/other_libs/mediatr.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+## Adding the event listener to your dependency injection container
+
+To start the listener we have to call the `StartAsync()` method on our `DiscordEventListener` class from inside of our main function. To do this, first register the `DiscordEventListener` class in your dependency injection container and get a reference to it in your main method.
+
+[!code-csharp[Starting the event listener](samples/MediatrStartListener.cs)]
+
+## Creating your notification handler
+
+MediatR publishes notifications to all of your notification handlers that are listening for a specific notification. We will create a handler for our newly created `MessageReceivedNotification` like this
```
Comment (csmir) [inline docs/guides/other_libs/mediatr.md]: 'this:'

File: docs/guides/other_libs/mediatr.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+## Creating notifications
+
+The way MediatR publishes events throughout your applications is through notifications and notification handlers. For this guide we will create a notification to handle the `MessageReceived` event on the `DiscordSocketClient`.
+
+[!code-csharp[Creating a notification](samples/MediatrCreatingMessageNotification.cs)]
+
+## Creating the notification publisher / event listener
+
+For MediatR to actually publish the events we need a way to listen for them. We will create a class to listen for discord events like so.
```
Comment (csmir) [inline docs/guides/other_libs/mediatr.md]: 'so:'

File: docs/guides/other_libs/mediatr.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+# Configuring MediatR
+
+## Prerequisites
+
+- A simple bot with dependency injection configured
+
+## Downloading the required packages
+
+You can install the following packages through your IDE or go to the nuget link to grab the dotnet cli command.
```
Comment (csmir) [inline docs/guides/other_libs/mediatr.md]: 'NuGet'

**Review by csmir — CHANGES_REQUESTED** [review]

**Comment by dukesteen** [comment]
Fixed in [b586319](https://github.com/discord-net/Discord.Net/pull/2218/commits/b58631905c1f03f99c55edccaae4770cf3ccead8)

**Review by csmir — APPROVED** [review]
lgtm

**Review by quinchs — APPROVED** [review]

**Pull request merged**